pkg/netutils/iface.go

```diff
@@ -0,0 +1,127 @@
+package netutils
+
+import (
+	"errors"
+	"net"
+	"syscall"
+
+	"github.com/vishvananda/netlink"
+)
+
+func GetDefaultGatewayInterface() (*net.Interface, error) {
+	routes, err := netlink.RouteList(nil, syscall.AF_INET)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, route := range routes {
+		if route.Dst == nil || route.Dst.String() == "0.0.0.0/0" {
+			if route.LinkIndex <= 0 {
+				return nil, errors.New("found default route but could not determine interface")
+			}
+			return net.InterfaceByIndex(route.LinkIndex)
+		}
+	}
+
+	return nil, errors.New("unable to find default route")
+}
+
+func GetDefaultV6GatewayInterface() (*net.Interface, error) {
+	routes, err := netlink.RouteList(nil, syscall.AF_INET6)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, route := range routes {
+		if route.Dst == nil || route.Dst.String() == "::/0" {
+			if route.LinkIndex <= 0 {
+				return nil, errors.New("found default v6 route but could not determine interface")
+			}
+			return net.InterfaceByIndex(route.LinkIndex)
+		}
+	}
+
+	return nil, errors.New("unable to find default v6 route")
+}
+
+func GetInterfaceByIP(ip net.IP) (*net.Interface, error) {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, iface := range ifaces {
+		err := ifaceIPv4AddrMatch(&iface, ip)
+		if err == nil {
+			return &iface, nil
+		}
+	}
+
+	return nil, errors.New("no interface with given IP address found")
+}
+
+func GetInterfaceByIPv6(ip net.IP) (*net.Interface, error) {
```

Comment:
Since we are exporting this methods could we add a comment on top to document them?
Also, do we need to export? Do we want someone using kurl a lib re-using this code? if not, could we no longer export things that we do not want anyone using?